extra_data/keydata.py

```diff
@@ -10,6 +10,119 @@
+    if not isinstance(indexing, tuple):
+        indexing = (indexing,)
+
+    ndim = len(shape)
+
+    # expanding Ellipsis
+    if any(x is Ellipsis for x in indexing):
+        ellipsis_idx = indexing.index(Ellipsis)
+        # Count non-Ellipsis, non-newaxis dimensions
+        n_before = sum(1 for x in indexing[:ellipsis_idx] if x is not None)
+        n_after = sum(1 for x in indexing[ellipsis_idx + 1:] if x is not None)
+        n_ellipsis = max(0, ndim - n_before - n_after)
+
+        # Replace Ellipsis with appropriate number of colons
+        indexing = (indexing[:ellipsis_idx] + 
+                   (slice(None),) * n_ellipsis + 
+                   indexing[ellipsis_idx + 1:])
+
+    # Separate newaxis from array dimensions
+    array_indices = []
+    new_axes_indices = []
+    dim_idx = 0
+
+    for position, idx in enumerate(indexing):
+        if idx is None:
+            new_axes_indices.append(position)
+        else:
+            if dim_idx < ndim:
+                array_indices.append(idx)
+                dim_idx += 1
+
+    # Pad with slice(None) if indexing is shorter than ndim
+    while len(array_indices) < ndim:
+        array_indices.append(slice(None))
+
+    # Process each index and expand coordinates
+    result = []
+
+    for idx, size in zip(array_indices, shape):
+        if isinstance(idx, slice):
+            # slice
+            if idx == slice(None) and not expand_full:
+                result.append(None)
+            else:
+                start, stop, step = idx.indices(size)
+                result.append(np.arange(start, stop, step))
+
+        elif isinstance(idx, int):
+            # single integer
+            normalized_idx = idx if idx >= 0 else size + idx
+            result.append(normalized_idx)
+
+        elif isinstance(idx, np.ndarray) and idx.dtype == bool:
+            # boolean array
+            if idx.shape[0] != size:
+                raise IndexError(f"Boolean index size {idx.shape[0]} doesn't match dimension size {size}")
+            result.append(np.where(idx)[0])
+
+        elif isinstance(idx, (list, tuple)):
+            # Check if it's a boolean list
+            if all(isinstance(x, (bool, np.bool_)) for x in idx):
+                # Boolean list
+                if len(idx) != size:
+                    raise IndexError(f"Boolean index size {len(idx)} doesn't match dimension size {size}")
+                result.append(np.where(np.array(idx))[0])
+            else:
+                # Integer list/tuple
+                arr = np.asarray(idx)
+                # Normalize negative indices
+                arr = np.where(arr < 0, arr + size, arr)
+                result.append(arr)
+
+        elif isinstance(idx, np.ndarray):
+            # Handle integer array
+            arr = idx.copy()
+            # Normalize negative indices
+            arr = np.where(arr < 0, arr + size, arr)
+            result.append(arr)
+
+        else:
+            # Default: all indices
```

Comment:
Do we know what remaining cases this `else` covers? It seems odd to assume anything that's not covered above is equivalent to `slice(None)`.